src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,191 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final int HALF_OF_SIZE = 2;
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> previous;
+
+        public Node(Node<T> previous, T item, Node<T> next) {
+            this.previous = previous;
+            this.item = item;
+            this.next = next;
+        }
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        return addLast(value);
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index <= size && index >= 0) {
+            if (index == 0) {
+                addFirst(value);
+            } else if (index == size) {
+                addLast(value);
+            } else {
+                insert(getByIndex(index), value);
+            }
+        } else {
+            throw new IndexOutOfBoundsException("Index is out of bounds");
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T element : list) {
+            addLast(element);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < size && index >= 0) {
+            return getByIndex(index).item;
+        }
+        throw new IndexOutOfBoundsException("Index is out of bounds");
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T previousValue;
+        if (index < size && index >= 0) {
+            previousValue = getByIndex(index).item;
+            getByIndex(index).item = value;
+            return previousValue;
+        }
+        throw new IndexOutOfBoundsException("Index is out of bounds");
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < size && index >= 0) {
+            T removedElement;
+            if (index == 0) {
+                removedElement = head.item;
+                removeFirst();
+                return removedElement;
+            } else if (index == size - 1) {
+                removedElement = tail.item;
+                removeLast();
+                return removedElement;
```

Comment:
These index checks can be done in removeElement() method so we don't need removeFirst() and removeLast();